From de353fb99b35ee15bc652c217a7475590ac56612 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Markus=20M=C3=BCtzel?= <markus.muetzel@gmx.de>
Date: Sat, 5 Sep 2026 18:42:19 +0200
Subject: [PATCH] Support optional installation of parpack in CMake config
 files

Some distributions (e.g., Debian/Ubuntu or MSYS2) split the parpack library into
a separate package.
That leads to issues if their arpack package is installed but their parpack
package is not.

Auto-generated CMake target files check whether the respective files actually
exist on disc. Because both the arpack import target and the parpack import
target are included in the same auto-generated CMake target file of arpack-ng
(if it was configured to be built parpack), that leads to a fatal error when
attempting to import the arpack CMake target in downstream projects while the
parpack files are not installed.

Support using the CMake config files in the case where distributions split
parpack into a separate package, and that package is not installed.
That is done by auto-generating a separate CMake target file for parpack,
and checking whether it exists before including it in the CMake config file.
Distributors can then move the CMake target file for parpack to their
separate package.
---
 CHANGES                        | 1 +
 CMakeLists.txt                 | 7 ++++++-
 cmake/arpackng-config.cmake.in | 6 +++++-
 3 files changed, 12 insertions(+), 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index d15ff0ecf..7a3273f83 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -860,7 +860,7 @@ install(FILES "${PROJECT_BINARY_DIR}/SRC/arpack${LIBSUFFIX}${ITF64SUFFIX}.pc"
 
 if (MPI)
   install(TARGETS parpack
-      EXPORT arpackngTargets
+      EXPORT parpackngTargets
       ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
       LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
       RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
@@ -905,6 +905,11 @@ endif()
 install(EXPORT arpackngTargets
   DESTINATION "${ARPACK_INSTALL_CMAKEDIR}"
 )
+if (MPI)
+  install(EXPORT parpackngTargets
+    DESTINATION "${ARPACK_INSTALL_CMAKEDIR}"
+  )
+endif()
 # Provide find_package for arpack-ng to users.
 configure_file(cmake/arpackng-config.cmake.in "${PROJECT_BINARY_DIR}/arpackng-config.cmake" @ONLY)
 configure_file(cmake/arpackng-config-version.cmake.in "${PROJECT_BINARY_DIR}/arpackng-config-version.cmake" @ONLY)
diff --git a/cmake/arpackng-config.cmake.in b/cmake/arpackng-config.cmake.in
index bafa3fed1..968ff6f4c 100644
--- a/cmake/arpackng-config.cmake.in
+++ b/cmake/arpackng-config.cmake.in
@@ -24,7 +24,7 @@ if (NOT @BUILD_SHARED_LIBS@)
 	if (@ICB@)
 		enable_language(Fortran)
 	endif()
-	if (@MPI@)
+	if (EXISTS "${CMAKE_CURRENT_LIST_DIR}/parpackngTargets.cmake")
 		include(FindMPI)
 		if (NOT TARGET MPI::Fortran)
 			find_dependency(MPI REQUIRED COMPONENTS Fortran)
@@ -34,6 +34,10 @@ endif()
 
 include("${CMAKE_CURRENT_LIST_DIR}/arpackngTargets.cmake")
 
+if(EXISTS "${CMAKE_CURRENT_LIST_DIR}/parpackngTargets.cmake")
+	include("${CMAKE_CURRENT_LIST_DIR}/parpackngTargets.cmake")
+endif()
+
 add_library(ARPACK::ARPACK ALIAS arpack)
 if (TARGET parpack)
 	add_library(PARPACK::PARPACK ALIAS parpack)
